tree-sitter-stack-graphs/src/functions.rs

```diff
@@ -0,0 +1,194 @@
+    functions.add("path-split".into(), path::PathSplit);
+}
+
+pub mod path {
+    use std::path::Component;
+    use std::path::PathBuf;
+    use tree_sitter_graph::functions::Function;
+    use tree_sitter_graph::functions::Parameters;
+    use tree_sitter_graph::graph::Graph;
+    use tree_sitter_graph::graph::Value;
+    use tree_sitter_graph::ExecutionError;
+
+    pub struct PathDir;
+    pub struct PathFileExt;
+    pub struct PathFileName;
+    pub struct PathFileStem;
+    pub struct PathJoin;
+    pub struct PathNormalize;
+    pub struct PathSplit;
+
+    impl Function for PathDir {
+        fn call(
+            &mut self,
+            _graph: &mut Graph,
+            _source: &str,
+            parameters: &mut dyn Parameters,
+        ) -> Result<Value, ExecutionError> {
+            let path = PathBuf::from(parameters.param()?.into_string()?);
+            parameters.finish()?;
+
+            let path = path.parent();
+            Ok(path
+                .map(|p| p.to_str().unwrap().into())
+                .unwrap_or(Value::Null))
+        }
+    }
+
+    impl Function for PathFileExt {
+        fn call(
+            &mut self,
+            _graph: &mut Graph,
+            _source: &str,
+            parameters: &mut dyn Parameters,
+        ) -> Result<Value, ExecutionError> {
+            let path = PathBuf::from(parameters.param()?.into_string()?);
+            parameters.finish()?;
+
+            let path = path.extension();
+            Ok(path
+                .map(|p| p.to_str().unwrap().into())
+                .unwrap_or(Value::Null))
+        }
+    }
+
+    impl Function for PathFileName {
+        fn call(
+            &mut self,
+            _graph: &mut Graph,
+            _source: &str,
+            parameters: &mut dyn Parameters,
+        ) -> Result<Value, ExecutionError> {
+            let path = PathBuf::from(parameters.param()?.into_string()?);
+            parameters.finish()?;
+
+            let path = path.file_name();
+            Ok(path
+                .map(|p| p.to_str().unwrap().into())
+                .unwrap_or(Value::Null))
+        }
+    }
+
+    impl Function for PathFileStem {
+        fn call(
+            &mut self,
+            _graph: &mut Graph,
+            _source: &str,
+            parameters: &mut dyn Parameters,
+        ) -> Result<Value, ExecutionError> {
+            let path = PathBuf::from(parameters.param()?.into_string()?);
+            parameters.finish()?;
+
+            let path = path.file_stem();
+            Ok(path
+                .map(|p| p.to_str().unwrap().into())
+                .unwrap_or(Value::Null))
+        }
+    }
+
+    impl Function for PathJoin {
+        fn call(
+            &mut self,
+            _graph: &mut Graph,
+            _source: &str,
+            parameters: &mut dyn Parameters,
+        ) -> Result<Value, ExecutionError> {
+            let mut path = PathBuf::new();
+            while let Ok(component) = parameters.param() {
```

Comment:
Do we need to call `parameters.finish()?` after this?

Reply:
That is not necessary. `finish` only checks if all parameters were consumed, which is guaranteed already by the loop.